src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,454 @@
+
+	dma_sg_free(&config.elem_array);
+
+	dcache_invalidate_region((void *)dma_ext->dmabp.addr, size);
+
+	return 0;
+}
+
+
+
+/* TODO
+ * This is stub, that will be replaced when TLB support
+ * will be available.
+ */
+static int mm_map(uint32_t *address, uint32_t size, uint32_t attribs)
+{
+	uint32_t addr_align = HOST_PAGE_SIZE;
+	void *local_add;
+	/* allocate new buffer */
+	local_add = (void *)rballoc_align(0, SOF_MEM_CAPS_DMA,
+						size, addr_align);
+	tr_err(&lib_manager_tr, "mm_map() : pointer: 0x%x", local_add);
+	if (!address) {
+		tr_err(&lib_manager_tr, "dma_buffer_init(): alloc failed");
+		return -ENOMEM;
+	}
+
+	*address = (uint32_t)local_add;
+	return 0;
+}
+
+static int lib_manager_store_library(struct lib_manager_dma_ext *dma_ext, uint32_t lib_id,
+				uint32_t addr_align)
+{
+	uintptr_t buff_ptr = dma_ext->dmabp.addr;
+	uint32_t library_base_address;
+	struct sof_man_fw_desc *man_desc =
+			(struct sof_man_fw_desc *) (buff_ptr + SOF_MAN_ELF_TEXT_OFFSET);
+	uint32_t preload_size = man_desc->header.preload_page_count * HOST_PAGE_SIZE;
+	int ret = IPC4_FAILURE;
+
+	/* Prepare storage memory */
+	/* TODO: There is no memory map API yet, so simply allocate external MEMORY region. */
+	/*       Final solution will require allocation and mapping this memory */
+	/*       to address library_base_address = EXT_MEMORY_base + load_offset       */
+	/* library_base_address = (uint8_t *)load_offset; */
+	ret = mm_map((void *)&library_base_address, preload_size, 0);
+
+	tr_err(&lib_manager_tr, "mm_map() :%d, pointer: 0x%x", ret, library_base_address);
+	if (ret < 0)
+		return ret;
+
+	/* Copy data from temp_mft_buf to destination memory (pointed by library_base_address) */
+	memcpy_s((void *)library_base_address, CAVS18_MAX_MANIFEST_SIZE,
+			 (void *)buff_ptr, CAVS18_MAX_MANIFEST_SIZE);
+	dcache_writeback_invalidate_region((void *)library_base_address, CAVS18_MAX_MANIFEST_SIZE);
+	/* Now free temporary buffer */
+	ret = lib_manager_dma_buffer_free(&dma_ext->dmabp);
+	if (ret < 0)
+		return ret;
+
+	/* Copy remaining library part into storage buffer */
+	/* initialise the DMA buffer */
+	dma_ext->dmabp.size = preload_size - CAVS18_MAX_MANIFEST_SIZE;
+	dma_ext->dmabp.w_ptr = (uintptr_t)(library_base_address + CAVS18_MAX_MANIFEST_SIZE);
+	dma_ext->dmabp.r_ptr = (uintptr_t)(library_base_address + CAVS18_MAX_MANIFEST_SIZE);
+	dma_ext->dmabp.end_addr = dma_ext->dmabp.w_ptr + dma_ext->dmabp.size;
+	dma_ext->dmabp.addr = dma_ext->dmabp.r_ptr;
+
+	ret = lib_manager_load_data_from_host(dma_ext, (preload_size - CAVS18_MAX_MANIFEST_SIZE));
+	if (ret < 0)
+		return ret;
+
+	/* Now update sof context with new library */
+	ret = lib_manager_update_sof_ctx((struct sof_man_fw_desc *)library_base_address, lib_id);
+
+	return 0;
+}
+
+int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id)
+{
+	uint32_t addr_align;
+	int ret = IPC4_FAILURE;
+
+	lib_manager_init();
+	struct lib_manager_dma_ext dma_ext;
+
+	ret = lib_manager_dma_init(&dma_ext, dma_id);
```

Comment:
dont you need to check the return value?

Reply:
please check new flow